### PR TITLE
Pass through specific annotation to OCI runtime spec

### PR DESCRIPTION
**Pull request by jshachm**
**- What I did**
Before this patch, docker daemon doesn't pass any annotations
to OCI runtime spec.However annotations can be useful espacially
for Runtime (Virtual Machine): kata/runv/gvisor and so on.

It's not good enough to pass all labels as annotations for there
will be some labels hold just for images.So we just need to paas two
kinds of annotations from upper cluster like kubernets:

First is user specific annotations which with prefix of `annotations.`
The other is labels for docker used which is marked as internal.

Signed-off-by: Haomin <caihaomin@huawei.com>

**- How I did it**

@miaoyq, I pick up from #36181, and look into what `kubernetes` has done with `annotations`.
We try to separate `label` into different types and decide which can be paas through to `runtime`.

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/oci_linux.go [hunk 1]
```diff
@@ -911,6 +926,31 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
 		s.Linux.ReadonlyPaths = c.HostConfig.ReadonlyPaths
 	}
 
+	// We just need to passthrough two types of annotations to runtime
+	// first is user specific annotations which with prefix of `annotations.`
+	// the other is labels for docker used which is marked as internal.
+	for k, v := range c.Config.Labels {
+		logrus.Info(k)
+		if oldValue, ok := s.Annotations[k]; ok {
+			return nil, fmt.Errorf("Key %q already exisit in Annotations, the new value will be ignored,"+
```
Comment (arm64b) [inline daemon/oci_linux.go]: This function have only one return value, why we return 2 values here, am I missing sth?

File: daemon/oci_linux.go [hunk 2]
```diff
@@ -911,6 +926,31 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
...
+		// interl labels for docker
+		for _, internalKey := range internalLabelKeys {
+			if k == internalKey {
+				s.Annotations[k] = v
+				break
+			}
+		}
+	}
+
 	return &s, nil
```
Comment (arm64b) [inline daemon/oci_linux.go]: seems the code base of this RP is out of date

Comment (jshachm) [reply daemon/oci_linux.go]: I check the mainline and I think it's up to date...  ^_^

File: daemon/oci_linux.go (hunk 1 above)
Comment (jshachm) [reply daemon/oci_linux.go]: func returns `spec` and error

Comment (arm64b) [reply daemon/oci_linux.go]: https://github.com/moby/moby/blob/master/daemon/oci_linux.go#L37 , this one❓

Comment (jshachm) [reply daemon/oci_linux.go]: @arm64b I think we are talking about https://github.com/moby/moby/blob/c752b0991e31ba9869ab6a0661af57e9423874fb/daemon/oci_linux.go#L747

**Comment by jshachm** [comment]
ping @justincormack @thaJeztah @AkihiroSuda 

What do you think about the new way to pass through `annotations`

File: daemon/oci_linux.go (hunk 1 above)
Comment (arm64b) [reply daemon/oci_linux.go]: Ah, yes, you are right 😄

**Comment by thaJeztah** [comment]
Still hesitant to conflate labels and annotations through the same API

I want to bring this up again in a maintainers meeting, but that won't be this week due to DockerCon

**Comment by jshachm** [comment]
@thaJeztah  Thx very much！ Passing `annotation` using flag `label` is a compromise. It will be great if `docker` will support a flag like `annotation`.

\cc @WeiZhang555 @bergwolf @egernst @gnawux @sameo

File: daemon/oci_linux.go [hunk 3]
```diff
@@ -34,6 +34,21 @@ var (
 	deviceCgroupRuleRegex = regexp.MustCompile("^([acb]) ([0-9]+|\\*):([0-9]+|\\*) ([rwm]{1,3})$")
 )
 
+// These values are copy from k8s.io/kubernets/pkg/kubelet/dockershim types
```
Comment (kolyshkin) [inline daemon/oci_linux.go]: nit: `s/copy/copied/` or `s/copy/taken/`

File: daemon/oci_linux.go [hunk 4]
```diff
@@ -911,6 +926,31 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
...
+				"oldvalue is %q, new value is %q", k, oldValue, v)
+		}
+		// user specific annotations
+		if strings.HasPrefix(k, anntotaionPrefix) {
+			key := strings.TrimPrefix(k, anntotaionPrefix)
+			logrus.Info(key)
+			s.Annotations[key] = v
+		}
+
+		// interl labels for docker
```
Comment (kolyshkin) [inline daemon/oci_linux.go]: nit: `s/interl/internal/`

File: daemon/oci_linux.go [hunk 5]
```diff
@@ -911,6 +926,31 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
...
 	}
 
+	// We just need to passthrough two types of annotations to runtime
+	// first is user specific annotations which with prefix of `annotations.`
+	// the other is labels for docker used which is marked as internal.
+	for k, v := range c.Config.Labels {
+		logrus.Info(k)
+		if oldValue, ok := s.Annotations[k]; ok {
+			return nil, fmt.Errorf("Key %q already exisit in Annotations, the new value will be ignored,"+
+				"oldvalue is %q, new value is %q", k, oldValue, v)
```
Comment (kolyshkin) [inline daemon/oci_linux.go]: nit: `s/oldvalue/old value/`

**Comment by olljanat** [comment]
Derek add label: rebuild/*

**Comment by olljanat** [comment]
Derek add label: rebuild/janky

**Comment by cpuguy83** [comment]
I think it would be great to have a generic way to pass configuration down to a runtime, however it should be explicit rather than implicit (maybe some daemon level config, maybe a dedicated field).

Because this doesn't quite seem like the right approach and this has gone stale, I'm going to close, but this is not an outright rejection of the idea behind this.

**Pull request closed without merging**